### PR TITLE
Add new type of Exception to handle 429 Errors

### PR DESCRIPTION
**Pull request by gogasca**
Add new type of Exceptions

### CODE REVIEW

File: gcsfs/utils.py [hunk 1]
```diff
@@ -88,6 +88,19 @@ def read_block(f, offset, length, delimiter=None):
     return bytes
 
 
+class RateLimitException(Exception):
+    """Holds the message and code from cloud errors."""
+    def __init__(self, error_response=None):
+        if error_response:
```
Comment (martindurant) [inline gcsfs/utils.py]: Do you see both variants in practice?

File: gcsfs/utils.py [hunk 2]
```diff
@@ -119,4 +132,6 @@ def is_retriable(exception):
     errs += [str(e) for e in errs]
     if isinstance(exception, HtmlError):
         return exception.code in errs
+    if isinstance(exception, RateLimitException):
```
Comment (martindurant) [inline gcsfs/utils.py:132]: Can you please include a link that says this type of exception is retriable? If I were google, I could imagine not recommending to retry calls that are already over some rate quota.

File: gcsfs/utils.py (hunk 1 above)
Comment (gogasca) [reply gcsfs/utils.py]: Updated code. Just see one option

File: gcsfs/utils.py (hunk 2 above)
Comment (gogasca) [reply gcsfs/utils.py:132]: Done. I will retry as client (papermill). Do you suggest the client owns retry logic as this PR? Or gcsfs should do the retries with backoff? (I see some logic in core GCSFileSystem _call method.

Comment (martindurant) [reply gcsfs/utils.py:132]: It is correct that gcsfs should own the retries if we are convinced that the exception should indeed be retried in normal usage.

Comment (gogasca) [reply gcsfs/utils.py:132]: I agree with your initial statement, for 429 Exception, GCS system allows only writing/updating object once per second. In papermill we were are sending more updates than that..., let me know if this PR is good for now or any other changes are recommended. Also how often the gcsfs package gets updated? Thanks

Comment (martindurant) [reply gcsfs/utils.py:132]: I think you are right, and this work well from my point of view.

**Comment by martindurant** [comment]
There is no regular release cycle here, but it can be done whenever I have a little spare time. I'll try to fit one in.

**Pull request merged**